File: leetcode/graph/834. Sum of Distances in Tree/py/main.py
import collections

class Solution:
    def sumOfDistancesInTree(self, N: int, edges):
        def dfs(u, parent):
          for v in adj_list[u]:
            if v != parent:
              dfs(v, u)
              cnt_nodes[u] += cnt_nodes[v]
              sum_sub_tree[u] += cnt_nodes[v] + sum_sub_tree[v]


        def dfs2(u, parent):
          for v in adj_list[u]:
            if v != parent:
              sum_sub_tree[v] = sum_sub_tree[u] - cnt_nodes[v] + N - cnt_nodes[v]
              dfs2(v, u)

        adj_list = collections.defaultdict(list)
        cnt_nodes = [1] * N
        sum_sub_tree = [0] * N

        for u, v in edges:
          adj_list[u].append(v)
          adj_list[v].append(u)

        dfs(0, None)
        dfs2(0, None)

        return sum_sub_tree

File: leetcode/graph/834. Sum of Distances in Tree/js/main.js
// Time O(N)
// Space O(N)
const sumOfDistancesInTree = (N, edges) => {
  let adjList = new Map();
  let cntNodes = Array(N).fill(1);
  let sumSubTree = Array(N).fill(0);

  for (let i = 0; i < N; i++) {
    adjList.set(i, []);
  }

  for (let [u, v] of edges) {
    adjList.get(u).push(v);
    adjList.get(v).push(u);
  }

  dfs(0, null);

  dfs2(0, null);

  return sumSubTree;

  // sumDist(k) = sumDist(parent) - (number of nodes in subtree k) + (number of nodes outside subtree k)
  function dfs2(u, parent) {
    for (let v of adjList.get(u)) {
      if (v != parent) {
        sumSubTree[v] = sumSubTree[u] - cntNodes[v] + (N - cntNodes[v]);

        dfs2(v, u);
      }
    }
  }

  // получаем кол-во узлов для каждой ноды и сумму рутового поддерева
  function dfs(u, parent) {
    for (let v of adjList.get(u)) {
      if (v != parent) {
        dfs(v, u);

        cntNodes[u] += cntNodes[v];
        sumSubTree[u] += cntNodes[v] + sumSubTree[v];
      }
    }
  }
};
